[Frontend][AST][IRGen] Improve availability support.

al45tair opened the pull request:
Use .def files to generate code for feature availability and runtime version to OS version mappings.

Also add a min-runtime-version option that can be used to avoid problems when building on Linux and Windows where because the runtime isn't part of the OS, availability doesn't solve the problem of trying to build the compiler against an older runtime.

Added some functions to IRGen to help with that problem; essentially we test *both* the deployment target availability *and* the new runtime version based availability.

rdar://121522431

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test macOS platform

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

eeckstein reviewed (approved):
Very nice, I like this!

Do I understand correctly that this PR contains of two separate features:
1. the new mechanism in the compiler for availability checking
2. passing the runtime version via build-script when building the compiler

(maybe worth splitting in two commits)

mikeash reviewed (approved):
Those .def files make everything so much nicer now.

al45tair commented:
> Do I understand correctly that this PR contains of two separate features:
> 
> 1. the new mechanism in the compiler for availability checking
> 2. passing the runtime version via build-script when building the compiler

Yes. I suppose there are logically three pieces:

1. The new .def-file based availability stuff.
2. The `-min-runtime-version` compiler flag (plus the new functions in IRGen that are enabled by (1) that check it).
3. The use of (2) in the `SwiftCompilerSources` `CMakeLists.txt`.

I guess I could split it into three commits maybe to cover those?

al45tair committed:
[AST] Improve availability support.

Use `.def` files to generate code for feature availability and runtime
version to OS version mappings.

rdar://121522431

al45tair committed:
[AST][IRGen] Add -min-runtime-version option and IRGen support code.

Add a `-min-runtime-version` option that can be used to avoid problems
when building on Linux and Windows where because the runtime isn't
part of the OS, availability doesn't solve the problem of trying to
build the compiler against an older runtime.

Also add functions to IRGen to make it easy to test feature
availability using both the runtime version and the existing Darwin
availability support.

rdar://121522431

al45tair committed:
[SwiftCompilerSources] Enable use of the `-min-runtime-version` option.

Turn on the use of `-min-runtime-version` when building the compiler.

rdar://121522431

al45tair commented:
OK, I've broken it up into three separate commits.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
I've also just realised that I somehow accidentally removed the code owners from the review list. I don't believe I did that myself, mind — it happened when I promoted this PR from a draft. Given that I'm tinkering with the availability code here, it's worth anyone interested taking a glance.

tshortli reviewed (commented):
Nice, this seems like a big improvement.

al45tair committed:
[AST] Update RuntimeVersions.def for 5.10.

This was just updated in #71297.

rdar://121522431

al45tair commented on include/swift/AST/RuntimeVersions.def:
```diff
@@ -0,0 +1,151 @@
+
+RUNTIME_VERSION(
+  (5, 10),
+  FUTURE
```
Updated this PR to match.

al45tair commented:
@swift-ci Please smoke test
